icu4j/main/classes/core/src/com/ibm/icu/impl/units/MeasureUnitImpl.java

```diff
@@ -0,0 +1,741 @@
+// © 2020 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html#License
+/*
+ *******************************************************************************
+ * Copyright (C) 2004-2020, Google Inc, International Business Machines
+ * Corporation and others. All Rights Reserved.
+ *******************************************************************************
+ */
+
+package com.ibm.icu.impl.units;
+
+import com.ibm.icu.impl.Assert;
```

Comment:
Do note: looks like Assert.assrt() runs for production code too, whereas the "assert" statement is debug-only.

That makes the "assert" more like U_ASSERT in C++, but might we in some cases want to have the exception thrown in production code? (I've not properly surveyed how assert / Assert.assrt is used in other ICU4J code.)

Reply:
Assert.assrt being this:

    public static void assrt(boolean val) {
        if (!val) throw new IllegalStateException("assert failed");
    }

Reply:
You should use the right exception for the job. Sometimes it will be IllegalArgumentException and sometimes IllegalStateException, and sometimes something else. For example it's rather confusing to get an IllegalArgumentException from a method that accepts no arguments.

As for "assert" (Java keyword) my policy is "never ever use it". It's banned inside Google for good reason (it has been the cause of real outages). It's code which is only compiled into the binary when you include debug information. Thus when it's used, it's quite possible for two people (compiling with different settings) to get binaries which behave differently.

assert also has the problem that it throws AssertionError, so hitting it is a hard fail for you library and the application that uses it, but in code compiled without debug info it means your library will continue in an undefined state potentially resulting in user data corruption.

If you have truly unreachable code, where you want a hard fail, I'd suggested just unconditionally throwing AssertionError directly, so it's in all versions of the code (but this should be vanishingly rare and only used for things like marking unreachable code when the compiler can't reason about it). Most things should be addressed by runtime exceptions like IllegalArgumentException or IllegalStateException.